src/IteratorImpl.hpp

```diff
@@ -75,6 +78,79 @@ class BorrowingIterator : public IteratorBase {
     Iter iter;
 };
 
+template<class Container, IteratorType type, class ValueType>
+class ValueOwningIterator : public IteratorBase {
+public:
+    typedef typename Container::const_iterator        Iter;
+    typedef typename CastFwd<ValueType>::Type::RsType ToValueType;
+
+    ValueOwningIterator(Container container)
+        : container_(std::move(container)), iter(container_.begin())
+    {
+        static_assert(type != itUnknown, "Iterator type must be valid");
+        update_current();
+    }
+
+    virtual IteratorType id() const override { return type; }
+    virtual bool valid() const override { return iter != container_.end(); }
+    virtual void* get() const override { return const_cast<ToValueType*>(&current); }
+
+    virtual void next() override {
+        if (valid()) {
+            ++iter;
+            update_current();
+        }
+    }
+
+private:
+
+    void update_current() {
+        if (valid()) {
+            current = cast_from_bw(*iter);
+        }
+    }
+
+    Container container_;
+    Iter iter;
+    ToValueType current;
```

Comment:
Field naming should be consistent. If there are ones with `_` then all fields should be named in the same fashion.